### DISCUSSION TITLE
Open existing note instead of template ?

### DISCUSSION THREAD

**Discussion by MichaVDM** [SilentVoid13/Templater#1343] [Help]
Hello,

I'm trying to use a new prompt template as below. Everything is quite simple, a prompt appears and lets me select which template I want to use. What I'm trying to achieve, is when I don't want to use a specific template at the time of a new file creation, to just open an existing "Workbox.md" file at the root of my vault (the XXX after the else statement in the code):

```
<%-*  let filetype = await tp.system.suggester(["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], ["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], false, "Which template do you want to use?") -%>
<%-* if (filetype === "Personal note") {  -%>
<%-tp.file.include("[[Personal note]]")-%>
<%-* } else if (filetype === "Book note") {  -%>
<%-tp.file.include("[[Book notes]]")-%>
<%-* } else if (filetype === "Unsorted note") {  -%>
<%-tp.file.include("[[Unsorted note]]")-%>
<%-* } else if (filetype === "Idée générale") {  -%>
<%-tp.file.include("[[Idée générale]]")-%>
<%-* } else if (filetype === "Readwise highlight") {  -%>
<%-tp.file.include("[[readwise]]")-%>
<%-* } else if (filetype === "PhD Research") {  -%>
<%-tp.file.include("[[PhD Research]]")-%>
<%-* } else { -%>
XXX
<%-* } -%>
```

Could anyone help, please? Thanks in advance!

**Comment by Zachatoo** [comment]
Here's a snippet for opening a file in the current tab.

```js
app.workspace.getLeaf().openFile(tp.file.find_tfile("Workbox"));
```

One additional consideration. Templater has already created the file by the time your code is run, so you may want to consider deleting the newly created file if you don't want to apply a template.

```js
await app.vault.delete(tp.file.find_tfile(tp.file.path(true)));
```

**Reply by MichaVDM** [reply]
Dear Zachatoo, thank you very much for your answer! Unfortunately, my coding skills are quite low, and I don't know precisely where to put the lines or how to organize them.

I started by inserting only the `app.workspace.getLeaf().openFile(tp.file.find_tfile("Workbox"));` snippet in the code above, and it fails in a weird manner, starting by opening the Workbox file, but immediately switching to a new 'untitled' file, with "undefined" as first and only word in the body of the text. The "good" news, is that there is no Templater error showing up, so it doesn't break.

This brings me to the rest of your suggestion about deletion, because I get the impression there could be something linked, in the way I unavoidably end up with a newly created file, instead of remaining in my Workbox file... But, when I try implementing your second line of code, I just break everything, because I don't know how to arrange the lines (sorry again, for my coding skills...)

Here's the code that doesn't produce errors but ends in the creation of a new untitled file:

`<%-*  let filetype = await tp.system.suggester(["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], ["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], false, "Which template do you want to use?") -%>
<%-* if (filetype === "Personal note") {  -%>
<%-tp.file.include("[[Personal note]]")-%>
<%-* } else if (filetype === "Book note") {  -%>
<%-tp.file.include("[[Book notes]]")-%>
<%-* } else if (filetype === "Unsorted note") {  -%>
<%-tp.file.include("[[Unsorted note]]")-%>
<%-* } else if (filetype === "Idée générale") {  -%>
<%-tp.file.include("[[Idée générale]]")-%>
<%-* } else if (filetype === "Readwise highlight") {  -%>
<%-tp.file.include("[[readwise]]")-%>
<%-* } else if (filetype === "PhD Research") {  -%>
<%-tp.file.include("[[PhD Research]]")-%>
<%-* } else { -%>
<%-app.workspace.getLeaf().openFile(tp.file.find_tfile("Workbox"))-%>
<%-* } -%>
`

And here's the code the doesn't work at all:

`<%-*  let filetype = await tp.system.suggester(["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], ["Personal note", "Book note", "Unsorted note", "Idée générale", "Readwise highlight", "PhD Research"], false, "Which template do you want to use?") -%>
<%-* if (filetype === "Personal note") {  -%>
<%-tp.file.include("[[Personal note]]")-%>
<%-* } else if (filetype === "Book note") {  -%>
<%-tp.file.include("[[Book notes]]")-%>
<%-* } else if (filetype === "Unsorted note") {  -%>
<%-tp.file.include("[[Unsorted note]]")-%>
<%-* } else if (filetype === "Idée générale") {  -%>
<%-tp.file.include("[[Idée générale]]")-%>
<%-* } else if (filetype === "Readwise highlight") {  -%>
<%-tp.file.include("[[readwise]]")-%>
<%-* } else if (filetype === "PhD Research") {  -%>
<%-tp.file.include("[[PhD Research]]")-%>
<%-* } else { -%>
<%-app.workspace.getLeaf().openFile(tp.file.find_tfile("Workbox"));
await app.vault.delete(tp.file.find_tfile(tp.file.path(true)))-%>
<%-* } -%>
`

So, I'm very sorry to ask, but could you rearrange the code for me in the example above? If it doesn't work, we'll eventually leave it there because I don't want you to lose your time, but I'm curious to know how to organize the lines in order for the code to function. Thank you very much in advance!

**Comment by kquinsland** [comment]
@MichaVDM I'm not 100% sure what it is that you're trying to do but I stumbled onto this thread because I had the same core question: how do switch to a file if it exists instead of trying to create it?

I agree that there's not a ton of intermediate documentation w/r/t Templater and using it with Obsidian so it can be tricky to figure out how to do some things beyond what's in the introduction / guide docs.

I _think_ the answer to your question is "how do I prevent the creation of the new `Untitled` note after `Workbox` is opened" and the answer is to just `return;` ASAP.

I have a generic `worklog` template that looks like this:

```md
<%*
// Omitted here: a bunch of JS code that builds up a list of `folders` from my vault.
// Basically I have hundreds of folders but am only interested in a sub-set of them so I have to list all the folders in my vault and then
// apply basic filter logic to eliminate some...

// Ask the user for the project or folder
const folder = await tp.system.suggester(folders, folders);

// Check if the user selected nothing (null)
if (folder === null) {
  // Stop execution if no selection is made; this prevents the Untitled note from being created
  return;
}

// omitted here: more code that's specific to my vault; basically it does some additional processing based on the user's
// selection to populate some variables that will be used in the note body.
// E.G: if the user selected X or Y from the list, put `kind/X` in the list of tags that will later make it into the front matter.



// Use Obsidian's API to check if the file already exists
const existingFile = tp.file.find_tfile(filePath);

if (existingFile) {
    console.debug(`file exists already. existingFile: ${existingFile}`)
    app.workspace.getLeaf().openFile(existingFile);
	// Stop execution NOW to prevent a file from being created regardless.
	return;
} else {
    // If the file doesn't exist, proceed to create or move the file as originally intended
    // Note: Since we're creating a new file, appending '.md' is necessary as Templater doesn't automatically add the extension
    console.log(`file does not exist already. filePath: ${filePath}`)
    await tp.file.move(filePath);

}

%>

// The rest of my note template here

```


For anybody else that stumbles on this thread, there's a few things I want to point out:

1. You can use `<% .... %>` to do multi-line templater expressions so you don't have to do `<%-* ...  -%>` on every line; this makes things a lot easier to read :).
2. If you call `return;` anywhere before `tp.file.{create|move...}` **then no file will be created**
3. **ChatGPT is actually pretty decent at putting together templater expressions**. I am a programmer but I don't work with JS much/at-all. CGPT is very good with javascript but it's not very good with templater / obsidian specific APIs so I got 90% of what I needed from chatGPT but then had to "patch up" it's solution

E.g.:

Here is the code that chatGPT suggested I use to check if a note exists or not:

```js
// Check if the file already exists
this.app.vault.getFiles().forEach(file => {
    if (file.path === `${filePath}.md`) {
        existingFilePath = file.path;
    }
});
```

While this _probably_ works, told chatGPT this:

````text
User
Instead of looping for each file to see if it exists, here is a more "obsidian" API to do so:


```
# This will return an object if the file can be found in the vault
const existing_file = tp.file.find_tfile(filePath)

# This will open a tab w/ the file
app.workspace.getLeaf().openFile(existing_file);
```
````

And it incorporated that into something that was more or less correct and did what I needed it to.


Hope that helps!

**Reply by MichaVDM** [reply]
Hi kquinsland, and thank you very much for your response! My problem is indeed close to what you describe and I think your help will be very useful in obtaining what I want. I just want to switch to the Workbox file if I don't select a template from the prompt.

I've tried playing around with the code without implementing too much of the code you provided, because I didn't want to select among existing projects, BUT, the magic happened when I thought of just trying your n°1 advice!!

> You can use <% .... %> to do multi-line templater expressions so you don't have to do <%-* ...  -%> on every line; this makes things a lot easier to read :).

I've removed the excessive <% .... %> and replaced the missing code (XXX) with "app.workspace.activeLeaf.openFile(tp.file.find_tfile("Workbox"));" and now it works like a charm, THANK YOU VERY MUCH!

Eventually, one detail, the 'getLeaf' instead of 'openLeaf' chatGPT seems to suggest, doesn't work...

Oh, and a new problem came up, just in case you can imagine a quick fix. When I don't select a template, it opens the Workbox file, perfect. But when I do select a template, it doesn't run the complete code. It creates and names the file correctly but doesn't insert the info prompted, while the template works fine when I directly select the same template... Could it be something like inserting time? An await command of some sort I should add?

Thanks again!

**Reply by kquinsland** [reply]
You need to use `<% $yourJSExpressionHere %>` syntax to fill out the file.

I try to keep the code between `<% %>` simple and to the point for readability to almost all of my templates have between 10-100 lines of JS in the "header" and just direct references to a variable in the `<%%>` block where I want content to be printed.

```md
<%*
 // Lots of JS here:
let foo = someJsExpression;
let bar = foo == none? "uh oh!" : "No problem";
%>

The intermeidate was: <% foo %> with result <% bar %>

// more md content here ...

```

**Comment by MichaVDM** [comment]
Thanks a lot to @kquinsland and @Zachatoo for your precious help! I think I can manage again from here. I don't know precisely how the GitHub "conversation" works, so this is not an answer in form of a solution, but the info shared by both can definitely be of some help to others!
